website/registrations/utils.py

```diff
@@ -0,0 +1,349 @@
+                                   'type': question.get('type', 'string'),
+                                   'options': question.get('options', [])}
+        for page in registration_schema.schema['pages']:
+            for question in page['questions']:
+                schema_questions[question['qid']] = get_question_validations(question)
+                if 'properties' in question:
+                    for nested_question in question['properties']:
+                        qid = '{}{}'.format(question['qid'], nested_question['id'])
+                        schema_questions[qid] = get_question_validations(nested_question)
+        return schema_questions
+
+    def log_error(self, **kwargs):
+        self.errors.append({
+            'header': kwargs['header'],
+            # TODO: add a method to generate human friendly excel column index
+            'column_index': kwargs['column_index'],
+            'row_index': kwargs['row_index'],
+            'missing': kwargs.get('missing', False),
+            'invalid': kwargs.get('invalid', False)
+        })
+
+    def validate_csv_header_list(self):
+        expected_headers = self.validations.keys()
+        actual_headers = self.headers
+        diff = set(expected_headers) - set(actual_headers)
+        if len(diff):
+            raise ValidationError('Invalid csv headers: {}'.format(','.join(diff)))
+
+    def get_parsed(self):
+        parsed = []
+        for row in self.rows:
+            parsed.append({'csv_raw': row.get_raw_value(), 'csv_parsed': row.get_parsed_value()})
+        return {'schema_id': self.schema_id, 'registrations': parsed}
+
+    def validate(self):
+        self.validate_csv_header_list()
+        for row in self.rows:
+            row.validate()
+
+class Row():
+    def __init__(self, row_dict, validations, log_error):
+        self.row_dict = row_dict
+        self.cells = [Cell(header.lower(),
+                           value, validations[header.lower()],
+                           functools.partial(log_error, column_index=column_index))
+                           for column_index, (header, value) in enumerate(row_dict.items())]
+
+    def to_json(self):
+        json_value = {}
+        for cell in self.cells:
+            json_value.update(cell.get_parsed_value())
+        return json.dumps(json_value)
+
+    def get_metadata(self):
+        parsed_metadata = {}
+        for cell in self.cells:
+            if cell.is_metadata:
+                parsed_metadata.update(cell.get_parsed_value())
+        return parsed_metadata
+
+    def get_registration_responses(self):
+        parsed_responses = {}
+        for cell in self.cells:
+            if not cell.is_metadata:
+                parsed_responses.update(cell.get_parsed_value())
+        return parsed_responses
+
+    def get_parsed_value(self):
+        return {'metadata': self.get_metadata(),
+                'registration_responses': self.get_registration_responses()}
+
+    def get_raw_value(self):
```

Comment:
Would the `DictReader` for csv remove `"` or `'` or any other chars that would be in a raw CSV row?

Reply:
for the purposes of meeting uniqueness constraint on each row, it feels like cleaning the raw string feels like an overkill

Reply:
My thoughts exactly. So if the `DictReader` strips the quotation marks from the values and/or do other encoding on special chars, then the `get_raw_value` function is not really constructing a string that matches the raw value, because the quotation marks and/or other special chars will be missing/different.

Reply:
Yep, the parser can't give us the raw. So one work-around is we do a second pass to read one line at time. Need to make sure the one line matches the `parsed_value`.

Reply:
It's not very easy or efficient to do a second pass though. I am starting to wonder whether keeping a `raw` value is even necessary.